### PR TITLE
Stop http2 test server with error code when failures occur

### PR DESCRIPTION
**Pull request by ericgribkoff**
The `twisted` module uses deferred errors rather than standard Python exceptions. This extends the http2 test server to register a callback for errors when binding the server to a port, and causes the server to exit with a non-zero status code when this occurs.

I can't reproduce the failures in https://github.com/grpc/grpc/issues/10056 locally, but this is potentially related - I still need to look and see if the docker servers are sharing a network space, in which case other servers on ports 8080-8085 would silently block the http2 server from doing anything. 

Regardless of whether this relates to https://github.com/grpc/grpc/issues/10056, this is desirable behavior for the http2 server in the case of future errors.

### CODE REVIEW

File: test/http2_test/http2_test_server.py [hunk 1]
```diff
@@ -83,6 +86,17 @@ def parse_arguments():
     )
   return parser.parse_args()
 
+def listen(endpoint, test_case):
+  deferred = endpoint.listen(H2Factory(test_case))
+  def listen_error(reason):
+    # If listening fails, we stop the reactor and exit the program
+    # with exit_code = 1.
+    global _exit_code
+    exit_code = 1
```
Comment (ncteisen) [inline test/http2_test/http2_test_server.py]: Missing a preceding underscore here

File: tools/run_tests/run_interop_tests.py [hunk 2]
```diff
@@ -635,10 +635,7 @@ def cloud_to_cloud_jobspec(language, test_case, server_name, server_host,
       '--server_host=%s' % server_host,
   ]
   if test_case in _HTTP2_BADSERVER_TEST_CASES:
-    # We are running the http2_badserver_interop test. Adjust command line accordingly.
-    offset = sorted(_HTTP2_BADSERVER_TEST_CASES).index(test_case)
-    client_options = common_options + ['--server_port=%s' %
-                                       (int(server_port)+offset)]
+    client_options = common_options + ['--server_port=%s' % server_port]
```
Comment (ncteisen) [inline tools/run_tests/run_interop_tests.py]: Nit, but now that we don't handle offset here, you could include server port in common_options again. Might as well for readability. Also would suggest hoisting client_options out of this conditional, and then just doing a += in the block for interop_only_options

**Comment by ericgribkoff** [comment]
@ncteisen I added a health check that waits and verifies that the http2 interop server is ready before continuing on with the interop tests. If the health check does not succeed after 15 seconds, an exception is thrown and the script exits. I think this is reasonable behavior, as it's the same approach taken in [`dockerjob.py#docker_mapped_port`](https://github.com/grpc/grpc/blob/master/tools/run_tests/python_utils/dockerjob.py#L58) when docker fails to return a port for the regular interop servers.

The health check is implemented using docker's health check mechanism. The documentation is at https://docs.docker.com/engine/reference/run/#healthcheck. Some of the command-line flags seemed counter-intuitive to me. `--health-retries` is the number of failures necessary to report the status as 'unhealthy', not the total number of retries that will be attempted. The value I chose was arbitrary, as the only thing that matters should be the timeout in `dockerjob.py`.

I also snuck in a small change that causes the `http2_badserver_interop` tests to still obey the `-l` languages flag. Previously it was impossible to run anything except all of the languages when running the http2 badserver tests. This should be an improvement for manual testing.

File: tools/run_tests/run_interop_tests.py [hunk 3]
```diff
@@ -677,23 +677,32 @@ def server_jobspec(language, docker_image, insecure=False, manual_cmd_log=None):
...
   if language.safename == 'http2':
     # we are running the http2 interop server. Open next N ports beginning
     # with the server port. These ports are used for http2 interop test
     # (one test case per port).
-    port_args = list(
+    docker_args += list(
         itertools.chain.from_iterable(('-p', str(_DEFAULT_SERVER_PORT + i))
                                       for i in range(
                                           len(_HTTP2_BADSERVER_TEST_CASES))))
+    docker_args += [
```
Comment (ncteisen) [inline tools/run_tests/run_interop_tests.py]: Looks really good in general. Would you just throw a comment in here about docker health check and why all these extra flags are needed? On first glance the relationship between --health-cmd and docker inspect is non intuitive.

Comment (ncteisen) [inline tools/run_tests/run_interop_tests.py]: Though if someone was reading through this script for the first time, this sections would probably be the least of their worry's....

**Comment by ctiller** [comment]
Is there an ETA on getting this in?

**Comment by ericgribkoff** [comment]
It's ready, but the change requires at least docker 1.12, and one of the Jenkins workers for interop tests only has docker version 1.11. @matt-kwong is going to upgrade the docker version later tonight, then it should be good to go.

**Comment by matt-kwong** [comment]
Just rebuilt the interop test on this PR here: https://grpc-testing.appspot.com/job/gRPC_interop_pull_requests/12336/. Docker version is 12.6; ping me if there is still an issue.

**Comment by ericgribkoff** [comment]
@matt-kwong Thanks for upgrading docker!

Known flakes:
https://github.com/grpc/grpc/issues/9562
https://github.com/grpc/grpc/issues/9975
https://github.com/grpc/grpc/issues/9994

**Pull request merged**